### DISCUSSION TITLE
Yesterday's update to `anyio` breaks startlette

### DISCUSSION THREAD

**Discussion by fernandosilveira** [encode/starlette#2261] [Potential Issue] [answered]
Yesterday [`anyio` updated to version `4.0.0`](https://github.com/agronholm/anyio/releases/tag/4.0.0)  and starlette [accepts anything `>=3.4.0`](https://github.com/encode/starlette/blob/master/pyproject.toml#L31).

Apparently, the latest `anyio` no longer has `start_blocking_portal`, thus breaking the test client (e.g., [here](https://github.com/encode/starlette/blob/cb6d284877cd359a21a0f31ebf8103a7b285a848/starlette/testclient.py#L416)).

I haven't looked deep enough into it but I suppose it's reasonable that `anyio` did not feel obliged to maintain backwards compatibility on a major version upgrade.

**Comment by Kludex** [comment]
Issue created for it: https://github.com/encode/starlette/issues/2262.

Thanks for reporting it.

**Comment by agronholm** [comment, marked as answer]
You say "breaks starlette" but offer no proof. I personally tested AnyIO 4 against starlette's test suite before making the release, and apart from one line coverage miss, everything worked fine.

Why do you claim that AnyIO no longer has `start_blocking_portal`? Sure, you have to import it from `anyio.from_thread` now, just like the test client code you linked to does!

**Reply by fernandosilveira** [reply]
Hmm, makes sense Alex. Feel free to close the issue. I'm having a situation in my CI where startlette hadn't been upgraded to import from `anyio.from_thread` yet.

**Reply by fernandosilveira** [reply]
I suppose past versions of starlette that didn't import from the correct place might be broken if they get the latest `anyio`, wrong?

**Reply by agronholm** [reply]
Yeah, so you have to upgrade both.

**Reply by fernandosilveira** [reply]
Unfortunately I can't upgrade to the latest starlette without further planning. I have a pinned fastapi which itself pins to the broken starlette version (0.20.4)

**Reply by agronholm** [reply]
Why did you update AnyIO then?

**Reply by fernandosilveira** [reply]
I didn't. One of my deps pulled the most recent anyio ([starlette==0.20.4, apparently](https://github.com/encode/starlette/blob/0.20.4/setup.py#L40))

**Reply by agronholm** [reply]
It looks like FastAPI needs to bump its starlette dependency then.

**Reply by agronholm** [reply]
Following up on this, it seems that the current FastAPI release pins Starlette 0.27.0, where the test client does in fact import `start_blocking_portal` correctly. The question is then, why are you only pinning FastAPI and not all transitive dependencies?

**Reply by fernandosilveira** [reply]
Like I said we have an earlier fastapi which pins starlette 0.20.4. We resolved our issue by pinning anyio 3.7. Unfortunately pinning every single transient dependency doesn't seems like a reasonable approach.

Anyway just wanted to reemphasize that we really love starlette and we're more than willing to deal with one of situations like this.

Cheers

**Reply by agronholm** [reply]
Locking down transitive dependencies is not only reasonable, it's *standard practice*. This is what tools like Poetry, pip-tools and Rye (off the top of my head) do. On the ECMAScript side, npm and yarn do the same thing and have worked this way for many, many years.

**Reply by agronholm** [reply]
My recommendation for you is to see if you can upgrade FastAPI to the latest release which appears to be mostly compatible. Meanwhile I'm trying to persuade the author to upgrade FastAPI's pinned Starlette dependency.

**Reply by VictorBancho** [reply]
Hi @agronholm, it seems the latest version of anyio (4.0.0) also enforces the ExceptionGroup paradigm (which can result in breaking changes). Since Starlette 0.19.0 (March last year) the starlette dependency for anyio has been set as >=3.4.0, <5.0.0 which covers two major versions. I assume it was set like this for dev purposes to integrate anyio 4.x?  I see your point about locking transitive dependencies, and is  certainly valid and something I will incorporate in future. 

I wanted to ask, is the ExceptionGroup enforcement a desired consequence of integrating with anyio 4.x? And if so, is this what Starlette deems the new best practice in exception handling?

**Reply by agronholm** [reply]
> I wanted to ask, is the ExceptionGroup enforcement a desired consequence of integrating with anyio 4.x? And if so, is this what Starlette deems the new best practice in exception handling?

The new semantics of task groups always raising exception groups from them is in line with how `asyncio.TaskGroup` works, and how Trio nurseries work when `strict_exception_groups=True` is set (which will be the default in the future). I cannot speak for the Starlette devs regarding their best practices.

**Reply by agronholm** [reply]
I added code in Starlette 0.31 (I think?) that extracts a single exception out of a single-exception group, so as to restore compatibility. But that doesn't help FastAPI users who are still stuck with an old Starlette version.

**Reply by Kludex** [reply]
The problem was that I changed the pinning on AnyIO to allow version 4 when it didn't have the ExceptionGroup changes. At the time I think the idea was to have a release soon, but things got postponed, and we didn't restrict the range.

There's nothing to do now. The recommendation is to always pin your transitive dependencies, and I'm sorry for the inconvenience.

Latest starlette works as expected.
